### PR TITLE
Ajc remove i7 requirement

### PR DESCRIPTION
**Pull request by ambrosejcarr**
This PR removes the hard requirement that the user pass an index fastq file. The index is not used in any downstream steps in the pipeline and is thought by many to be superfluous. Removing the requirement increases pipeline flexibility.

### CODE REVIEW

**Comment by ambrosejcarr** [comment]
@jsotobroad 

I'm running into the following error on this PR: 

`Unable to build WOM node for WdlWorkflowCall 'target': Unable to build WOM node for Scatter '$scatter_0': Unable to build WOM node for If '$if_0': Unable to build WOM node for WdlTaskCall 'attach_barcodes': Invalid indexing target. You cannot index a value of type 'Array[File]?'`

... can I not index an optional Array[File]? That seems like it more or less invalidates the usefulness of the object, so I assume I'm doing something wrong. Can you take a look?

**Comment by jsotobroad** [comment]
yup unfortunately you have to do `Array[File] make_non_optional = select_first([i1])` to get a non optional array, you can then index `make_non_optional`

**Comment by ambrosejcarr** [comment]
@jsotobroad could you do a real review of this? your suggestion fixed the problem. :)

File: library/tasks/Attach10xBarcodes.wdl [hunk 1]
```diff
@@ -3,26 +3,26 @@ task Attach10xBarcodes {
...
-      --i1 "${i1}" \
+      ${"--i1 " + i1} \
       --u2 "${u2}" \
       --output-bamfile barcoded.bam \
       --whitelist "${whitelist}"
   }
   
   runtime {
-    docker: "quay.io/humancellatlas/secondary-analysis-python3-scientific:0.1.5"
+    docker: "quay.io/humancellatlas/secondary-analysis-sctools:ajc-remove-i7-requirement-517"
```
[the remark this answers is not in the document]
Comment (jsotobroad) [reply library/tasks/Attach10xBarcodes.wdl]: you said at standup that it moved from secondary analysis to skylab i believe or vice versa?  We shoudl just make sure the new name makes sense

File: pipelines/optimus/Optimus.wdl [hunk 2]
```diff
@@ -12,34 +12,60 @@ import "AlignTagCorrectUmis.wdl" as AlignTagCorrectUmis
...
+        input:
+          r1 = r1[index],
+          i1 = non_optional_i1[index],
+          u2 = FastqToUBam.bam_output,
+          whitelist = whitelist
+      }
     }
 
+    # if the index is not passed, proceed without it.
+    if (!defined(i1)) {
```
Comment (jsotobroad) [inline pipelines/optimus/Optimus.wdl]: I couldn't think of a better way to do this 😢

File: library/tasks/Attach10xBarcodes.wdl [hunk 3]
```diff
@@ -3,26 +3,26 @@ task Attach10xBarcodes {
   # attaches barcodes found in r1 and i1
```
Comment (jsotobroad) [inline library/tasks/Attach10xBarcodes.wdl]: I see that `\n` at the top!

File: library/tasks/Attach10xBarcodes.wdl (hunk 1 above)
[the remark this answers is not in the document]
Comment (ambrosejcarr) [reply library/tasks/Attach10xBarcodes.wdl]: This is the name of the branch, which is used during testing. Given that tests have passed, I think the first step is to finalize the sctools PR (also linked in the ticket) and then push a version tag to that repo which will auto-build it. 

This should be incorporated before we merge.

File: pipelines/optimus/Optimus.wdl (hunk 2 above)
Comment (ambrosejcarr) [reply pipelines/optimus/Optimus.wdl]: 👍

File: test/optimus/pr/test_inputs.json [hunk 4]
```diff
@@ -3,18 +3,19 @@
...
-    ]
+  "TestOptimusPR.r1": [
+    "gs://broad-dsde-mint-dev-teststorage/10x/demo/fastqs/pbmc8k_S1_L007_R1_001.fastq.gz",
+    "gs://broad-dsde-mint-dev-teststorage/10x/demo/fastqs/pbmc8k_S1_L007_R1_001.fastq.gz"
   ],
+  "TestOptimusPR.r2": [
+    "gs://broad-dsde-mint-dev-teststorage/10x/demo/fastqs/pbmc8k_S1_L007_R2_001.fastq.gz",
+    "gs://broad-dsde-mint-dev-teststorage/10x/demo/fastqs/pbmc8k_S1_L007_R2_001.fastq.gz"
+  ],
+  "TestOptimusPR.i1": [
```
[the remark this answers is not in the document]
Comment (ambrosejcarr) [reply test/optimus/pr/test_inputs.json]: I'm going to push a commit wherein we remove those i1 file inputs and see where the PR test fails. I expect that the bam file should no longer look the same, but that the pipeline should complete execution and only the _checker_ should fail. 

If this occurs, I think we're good to merge.

File: library/tasks/Attach10xBarcodes.wdl (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsotobroad) [reply library/tasks/Attach10xBarcodes.wdl]: Gotcha thanks for the explanation

File: test/optimus/pr/test_optimus_PR.wdl [hunk 5]
```diff
@@ -13,7 +13,10 @@ workflow TestOptimusPR {
   String expected_picard_metrics_hash
 
   # Optimus inputs
-  Array[Array[File]] fastq_inputs  # array of arrays of matched fastqs [ [r1, r2, i1], ... ]
+  Array[File] r1
+  Array[File] r2
+  Array[File] i1
```
Comment (jsotobroad) [inline test/optimus/pr/test_optimus_PR.wdl]: I think you'll have to make this optional if you want to test not passing in an `i1` file.  Does our testing infrastructure allow for testing a workflow against multiple input jsons / is this something we even want to do?

Comment (ambrosejcarr) [reply test/optimus/pr/test_optimus_PR.wdl]: It should allow this, yes. I'm not sure we'd want to put in this use case.

**Comment by ambrosejcarr** [comment]
Good call, I will bring it up at stand-up tomorrow. 👍 

I doubt I'll merge this before then.

File: library/tasks/Attach10xBarcodes.wdl (hunk 1 above)
Comment (jishuxu) [inline library/tasks/Attach10xBarcodes.wdl]: I remember there was some conversation on docker tagging and docker auto-build from each branch. I don't think we settle on how to do this yet.

File: pipelines/optimus/Optimus.wdl (hunk 2 above)
Comment (jishuxu) [reply pipelines/optimus/Optimus.wdl]: i guess if we 100% sure we won't need I1,we can just simple take I1 out of the pipeline.

File: pipelines/optimus/Optimus.wdl [hunk 6]
```diff
@@ -12,34 +12,60 @@ import "AlignTagCorrectUmis.wdl" as AlignTagCorrectUmis
 # format.
 workflow Optimus {
 
-  Array[Array[File]] fastq_inputs  # array of arrays of matched fastqs [ [r1, r2, i1], ... ]
-  File whitelist  # 10x genomics cell barcode whitelist for 10x V2
+  # Sequencing data inputs (fastq)
+  Array[File] r1 # forward read, contains cell barcodes and molecule barcodes
```
Comment (jishuxu) [inline pipelines/optimus/Optimus.wdl]: Array[File] is much better!

**Comment by ambrosejcarr** [comment]
Sorry I guess I wasn't clear about that Jishu -- there are lots of cases wherein one might pass those files, they are just not _currently_ consumed by the pipeline in any way.

**Comment by jishuxu** [comment]
we might need to do some test with various input, such as w/o I1 reads. Just want to make sure pipeline run through, nothing fancy. As soon as these tests run through, it is fine with me to merge.

**Comment by ambrosejcarr** [comment]
jenkins retest

**Comment by ambrosejcarr** [comment]
@jsotobroad @jishuxu I've confirmed that making the i1 file optional and removing the i1 inputs runs through the Optimus workflow. See: 

https://cromwell.mint-dev.broadinstitute.org/api/workflows/v1/239f572c-7b8f-4920-8bd8-236f6dbbbd9b/metadata?expandSubWorkflows=false

Note that this is the PR workflow, so the overall workflow fails the checker task. However, "target" succeeds and the output looks good -- it's just missing i1 tags and therefore fails the md5. Given this, I think that reverting the input json to pass the i1 file will be good to merge.

**Pull request merged**